File: placementprepration/tree/boun.java
import java.awt.image.RescaleOp;
import java.util.*;
import javax.swing.event.AncestorEvent;
class TreeNode{
    int data;
    TreeNode left, right;
    TreeNode(int val){
        this.data=val;
        this.left=null;
        this.right=null;
    }
    public class Binarytraversal{
        static Scanner sc=new Scanner(System.in);
        public static boolean isLeaf(Node root){
            return (root!=null && root.left==null&& root.right==null);

        }
        public static void addleft(TreeNode root ,List<Integer>res){
            TreeNode temp=root.left;
            while(temp!=null){
                if(!isleaf(temp)){
                    res.add(temp.data);
                }
                if(temp.left!=null){
                    temp=temp.left;
                }
                else{
                    temp=temp.right;
                }
            }
        }
        public static void addright(TreeNode root, List<Integer>res){
            TreeNode temp=root.right;
            List<Integer>tempres=new ArrayList<>();
            while(temp!=null){
                if(!isLeaf(temp)){
                    tempres.add(temp.data);
                }
                if(temp.right!=null){
                    temp=temp.right;
                }
                else{
                    temp=temp.left;
                }
                Collections.reverse(tempres);
                res.addAll(tempres);
            }
        }
        public static void addleaf(TreeNode root, List<Integer>res){
            if(root==null){
                return;
            }
            if(isleaf(root)){
                res.add(root.data);
                return;
            }
            addleaf(root.left,res);
            addleaf(root.right,res);
        }
        public static List<Integer>bound(TreeNode root){
            List<Integer>ans=new ArrayList<>();
            if(root==null){
                return ans;
            }
            if(!isleaf(root)){
                ans.add(root.data);
            }
        }
    }
}

File: placementprepration/tree/boun.py
def is_leaf(node):
    return node and node.left is None and node.right is None


class TreeNode:
    def __init__(self, val):
        self.data = val
        self.left = None
        self.right = None


def add_left(root, res):
    temp = root.left
    while temp:
        if not is_leaf(temp):
            res.append(temp.data)
        temp = temp.left if temp.left else temp.right


def add_right(root, res):
    temp = root.right
    tmp_res = []
    while temp:
        if not is_leaf(temp):
            tmp_res.append(temp.data)
        temp = temp.right if temp.right else temp.left
    res.extend(reversed(tmp_res))


def add_leaves(root, res):
    if root is None:
        return
    if is_leaf(root):
        res.append(root.data)
        return
    add_leaves(root.left, res)
    add_leaves(root.right, res)


def boundary(root):
    res = []
    if root is None:
        return res
    if not is_leaf(root):
        res.append(root.data)
    add_left(root, res)
    add_leaves(root, res)
    add_right(root, res)
    return res


def main():
    # Small demo tree
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    print(boundary(root))


if __name__ == "__main__":
    main()
